feature(notifications): refactor notification system for improved usability

hypeJunction opened the pull request:
Replaces #9946 (github wouldn't let me reopen)

Improves usability of instant and subscription notifications:
- Adds "origin" to notification object parameters making it easier to distinguish between
  instant and subscription notifications
- Adds a high level "prepare, notification" plugin hook that can be used to filter
  both subscription and instant notifications.
  Normalizes parameters sent to the hook for both subscription types.
- Adds "format, notification:$method" hook, which can be used to format notification subject
  and/or body, to make it suitable for the delivery method. For example, plugins may strip
  HTML tags from notifcation body for plaintext emails, or wrap the body in an HTML template
  for HTML emails.
- Instant notifications are now sent by the notifications service triggering
  a set of hooks similar to subscriptions notifications.
- Help developers debug notifications by logging notifications and notification
  event results

hypeJunction committed:
feature(notifications): refactor notification system for improved
usability

Improves usability of instant and subscription notifications:

- Adds "origin" to notification object parameters making it easier to distinguish between
instant and subscription notifications

- Adds a high level "prepare, notification" plugin hook that can be used to filter
both subscription and instant notifications.
Normalizes parameters sent to the hook for both subscription types.

- Adds "format, notification:$method" hook, which can be used to format notification subject
and/or body, to make it suitable for the delivery method. For example, plugins may strip
HTML tags from notifcation body for plaintext emails, or wrap the body in an HTML template
for HTML emails.

- Instant notifications are now sent by the notifications service triggering
a set of hooks similar to subscriptions notifications.

- Help developers debug notifications by logging notifications and notification
event results

- Notification event serialization is now more reliable and tested.

- Adds API to retrieve registered notification methods

- Adds ElggUser::getNotificationSettings() and
  ElggUser::setNotificationSetting()

hypeJunction commented:
@juho-jaakkola @mrclay This is the last time I am touching this. Please review sooner or later and let's do something with it before it goes stale again.

mrclay commented:
LGTM. A little over my head, but seems like a major improvement with no obvious faults.

mrclay commented:
Eh, rebase and go for it.

juho-jaakkola commented:
Sorry for taking so long! I'm going to review the shit out of this today.

juho-jaakkola commented on engine/classes/Elgg/Notifications/Event.php:
```diff
@@ -67,30 +72,16 @@ public function getActor() {
 	 */
 	public function getActorGUID() {
-		return $this->actor_guid;
+		return $this->actor ? $this->actor->guid : 0;
```
In what case is actor not set here? What does the `0` GUID refer to?

hypeJunction replied:
Edge case: user account deleted/disabled before notification queue is processed

juho-jaakkola replied:
Mention that in the docblock.

juho-jaakkola commented on engine/classes/Elgg/Notifications/Event.php:
```diff
@@ -108,7 +99,28 @@ public function getAction() {
+	}
+
+	/**
+	 * Export
```
Add some documentation about why this method is needed / what it is used for.

hypeJunction replied:
Logging. Will add a description

juho-jaakkola commented on engine/classes/Elgg/Notifications/InstantNotificationEvent.php:
```diff
@@ -0,0 +1,119 @@
+			$this->actor = _elgg_services()->session->getLoggedInUser();
+		}
+
+		$this->action = $action ? : self::DEFAULT_ACTION_NAME;
```
Huh, I didn't know you could do that. Even though it's been there since PHP 5.3!

> Since PHP 5.3, it is possible to leave out the middle part of the ternary operator.

juho-jaakkola commented on engine/classes/Elgg/Notifications/InstantNotificationEvent.php:
```diff
@@ -0,0 +1,119 @@
+	 * @return int
+	 */
+	public function getActorGUID() {
+		return $this->actor ? $this->actor->guid : 0;
```
Here also: what makes actor empty?

hypeJunction replied:
Deleted account

juho-jaakkola commented on engine/classes/Elgg/Notifications/NotificationsService.php:
```diff
@@ -172,22 +187,25 @@ public function enqueueEvent($action, $type, $object) {
-	 * @param int $stopTime The Unix time to stop sending notifications
-	 * @return int The number of notification events handled
+	 * @param int  $stopTime The Unix time to stop sending notifications
+	 * @param bool $matrix   If true, will return delivery matrix instead of a notifications event count
```
What are the use cases for the matrix? I don't see it being used anywhere within the PR.

Without really knowing more, I'm wondering whether we should add `getDeliveryMatrix()` instead of having mixed return value. But that probably depends on the use cases?

hypeJunction replied:
Number of successful events tells us nothing. Matrix gives a better idea of what has been done.  I kept the old return to avoid rewriting tests.

juho-jaakkola replied:
That's not a use case. ;)

But no worries, this is not a blocker either.

juho-jaakkola commented on engine/classes/Elgg/Notifications/NotificationsService.php:
```diff
@@ -197,129 +215,304 @@ public function processQueue($stopTime) {
+				if (!in_array($object->access_id, [ACCESS_PUBLIC, ACCESS_LOGGED_IN]) && !has_access_to_entity($object, $recipient)) {
+					// In this context we want to make sure public and logged in access is considered accessible
+					// in spite of what access SQL may say: when running in cron we don't have a logged in user,
+					// in walled garden public access may not be in user's access list
```
I have trouble understanding what exactly happens here. Can you give some example use cases?

hypeJunction replied:
For some reason tests were failing on this. I will need to investigate further. Basically when we don't have a logged in user, access ids 1 and 2 are not in access array hence get_entity fails

juho-jaakkola commented on engine/classes/ElggUser.php:
```diff
@@ -667,4 +667,40 @@ public function setPassword($password) {
+	 * @param bool   $enabled Enabled or disabled
+	 * @return bool
+	 */
+	public function setNotificationSetting($method, $enabled = true) {
```
I'm not sure whether it's wise to add any more methods within `ElggUser`. The class is already doing more than it perhaps should. Better would be to have a separate subscriptions service to take care of it:

```php
$subscriptions->setPersonalSetting($user, $method, $enabled);
```

But this is not a blocker. We can always deprecate the function later if we want to separate the feature.

hypeJunction replied:
It makes testing a lot easier

juho-jaakkola commented on engine/classes/ElggUser.php:
```diff
@@ -667,4 +667,40 @@ public function setPassword($password) {
+	 * @return bool
+	 */
+	public function setNotificationSetting($method, $enabled = true) {
+		$this->{"notification:method:$method"} = (bool) $enabled;
```
Is there reason not to simply remove the metadata if the method is disabled?

juho-jaakkola replied:
Side note: we could simplify this in Elgg 3. See #10185

mrclay replied:
See #10186. We should cast to int or string.

juho-jaakkola commented on engine/lib/notification.php:
```diff
@@ -95,6 +95,15 @@ function elgg_register_notification_method($name) {
 
 /**
+ * Returns registered delivery methods for notifications
+ * @return array
```
Add an example return value.

juho-jaakkola commented on engine/lib/notification.php:
```diff
@@ -524,37 +475,22 @@ function notify_user($to, $from, $subject, $message, array $params = array(), $m
-	$user_guid = (int)$user_guid;
 
-	if ($user_guid == 0) {
+	elgg_deprecated_notice(__FUNCTION__ . ' has been derepcated by ElggUser::getNotificationSettings()', '2.3');
```
Typo

juho-jaakkola replied:
Is there a reason to deprecate this in the first place?

hypeJunction replied:
The return value is really annoying to work with

juho-jaakkola commented on engine/tests/phpunit/Elgg/Notifications/NotificationsServiceTestCase.php:
```diff
@@ -0,0 +1,929 @@
+		$this->session->removeLoggedInUser();
+	}
+
+	public function testStoppingEnqueueEvent() {
```
The more descriptive names, the better. This could be e.g. `testPluginHookHandlerCanStopEnqueueEvent()`.

hypeJunction replied:
Not my test. Copied from the old case.

juho-jaakkola replied:
No reason to make it better now. :)

hypeJunction replied:
I have done enough "better" with this PR, just want it out of my life

juho-jaakkola commented on engine/tests/phpunit/Elgg/Notifications/NotificationsServiceTestCase.php:
```diff
@@ -0,0 +1,929 @@
+		$this->assertEquals(0, $this->notifications->processQueue($this->time));
+	}
+
+	public function testCanUseEnqueueHookToPreventSubscriptionNotificationEventFromQueueing() {
```
Now here's an exhaustive test name. ;)

juho-jaakkola commented:
Unit test are failing locally, but the failures don't seem to be related to the PR.

juho-jaakkola commented:
Green light for merging latest on Monday. :+1:

hypeJunction committed:
perf(db): no longer queries DB when entity access is predictable

AccessCollections::hasAccessToEntity() no longer queries the DB to
determine access to entities with ACCESS_LOGGED_IN and ACCESS_PUBLIC,
as well as access by owners.

Refs #10133

hypeJunction commented:
@juho-jaakkola Addressed your comments
@mrclay Please take a look at the last commit. I can't make the DB query mocks work for these edge cases, so I think we should just avoid querying the DB when the accessibility is predictable.

mrclay commented:
All LGTM

hypeJunction commented:
Merging. Out of sight, out of heart.

jdalsem commented:
:cake:
